src/Data/LLVM/BitCode/Assert.hs

```diff
@@ -80,43 +79,37 @@ recordSizeIn record ns =
 ----------------------------------------------------------------
 -- ** Types
 
--- | Assert that this thing is a pointer, get the underlying type
+-- | Assert that this thing is a @'PtrTo' ty@ and return the underlying @ty@.
+--
+-- Think carefully before using this function, as it will not work as you would
+-- expect when the type is an opaque pointer.
+-- See @Note [Pointers and pointee types]@.
```

Comment:
Should we add a WARNING pragma here and below (category "llvm-pointers" so that it can be explicitly disabled but is enabled by default)?  

See: https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/pragmas.html#warning-and-deprecated-pragmas

Reply:
> Should we add a WARNING pragma here and below

I'm of two minds on this. While `elimPtrTo` is almost assuredly not what you want in an opaque pointer world, there are still some places where we need to use it in `llvm-pretty-bc-parser` for backwards compatibility with old LLVMs. As a result, we would have to disable deprecation warnings in `llvm-pretty-bc-parser` wholesale to avoid this. 

> category "llvm-pointers" so that it can be explicitly disabled but is enabled by default

The category feature is cool, and this would give us a way to avoid the disabling-warnings-wholesale problem mentioned above. Sadly, warning categories are a feature that is only available in GHC 9.8 (not yet released) or later. We could add a category, but it wouldn't be useful to us for some time.